### PR TITLE
Use nethost.dll to find dotnet in ANCM

### LINKED ISSUE #8838: React to hostfxr locating logic
https://github.com/dotnet/core-setup/pull/5522 just got merged. Let's try this out so we can drop a majority of our hostfxr searching logic.

Work items:
- Include `nethost.dll` as part of the ANCM msis/installers and available in the IISIntegration subfolder. This is actually fairly complicated because anywhere we bring `aspnetcorev2.dll`, we now need to bring `nethost.dll`
- Verify no scenarios are broken with new finding logic.

### PR DESCRIPTION
**Pull request by jkotalik**
Fixes https://github.com/aspnet/AspNetCore/issues/8838

### CODE REVIEW

File: eng/Version.Details.xml [hunk 1]
```diff
@@ -381,6 +381,11 @@
       <Uri>https://github.com/dotnet/core-setup</Uri>
       <Sha>63abc77da6d99470caa5bfa0465afe244105e595</Sha>
     </Dependency>
+    <Dependency Name="Microsoft.NETCore.DotNetAppHost" Version="">
```
Comment (natemcmaster) [inline eng/Version.Details.xml]:
```suggestion
    <Dependency Name="Microsoft.NETCore.DotNetAppHost" Version="3.0.0-preview6-27730-01">
```

**Review by natemcmaster — APPROVED** [review]
Changes LGTM. If it works, :shipit:

**Comment by jkotalik** [comment]
Got a slight block on https://github.com/dotnet/core-setup/pull/6672

**Comment by jkotalik** [comment]
@BrennanConroy this is now ready for review. No rush though, I know SignalR + Npm is higher priority.

File: eng/Version.Details.xml [hunk 2]
```diff
@@ -389,6 +389,10 @@
       <Uri>https://github.com/dotnet/core-setup</Uri>
       <Sha>094b6111d3d8c90d1111db4d73fd9faa755d9413</Sha>
     </Dependency>
+    <Dependency Name="runtime.win-x64.Microsoft.NETCore.DotNetAppHost" Version="3.0.0-preview7-27812-08" CoherentParentDependency="Microsoft.EntityFrameworkCore">
```
Comment (BrennanConroy) [inline eng/Version.Details.xml]: Shouldn't the `CoherentParentDependency` be "Microsoft.Extensions.Logging"

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/StartupTests.cs [hunk 3]
```diff
@@ -140,7 +135,6 @@ public async Task StartsWithDotnetInstallLocation(RuntimeArchitecture runtimeArc
                     // Verify that in this scenario dotnet.exe was found using InstallLocation lookup
```
Comment (BrennanConroy) [inline src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/StartupTests.cs]: Remove comment

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/StartupTests.cs [hunk 4]
```diff
@@ -44,8 +43,6 @@ public async Task ExpandEnvironmentVariableInWebConfig()
 
         [ConditionalTheory]
         [InlineData("bogus", "", @"Executable was not found at '.*?\\bogus.exe")]
-        [InlineData("c:\\random files\\dotnet.exe", "something.dll", @"Could not find dotnet.exe at '.*?\\dotnet.exe'")]
-        [InlineData(".\\dotnet.exe", "something.dll", @"Could not find dotnet.exe at '.*?\\.\\dotnet.exe'")]
```
Comment (BrennanConroy) [inline src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/StartupTests.cs]: These no longer valid?

File: src/Hosting/Server.IntegrationTesting/src/Common/DotNetCommands.cs [hunk 5]
```diff
@@ -46,8 +46,9 @@ public static string GetDotNetHome()
         public static string GetDotNetInstallDir(RuntimeArchitecture arch)
         {
             var dotnetDir = DotNetHome;
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && arch == RuntimeArchitecture.x86)
             {
+                // x86 dotnet is now in a subfolder under dotnet
```
Comment (BrennanConroy) [inline src/Hosting/Server.IntegrationTesting/src/Common/DotNetCommands.cs]: Isn't x64 also in a subfolder?

File: eng/Version.Details.xml (hunk 2 above)
Comment (jkotalik) [reply eng/Version.Details.xml]: yeah bad copy pasta.

File: src/Hosting/Server.IntegrationTesting/src/Common/DotNetCommands.cs (hunk 5 above)
Comment (jkotalik) [reply src/Hosting/Server.IntegrationTesting/src/Common/DotNetCommands.cs]: Nope. x86 is but not x64.

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/StartupTests.cs (hunk 4 above)
Comment (jkotalik) [reply src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/StartupTests.cs]: I was lazy here I think. The input is invalid and should be tested but the error message won't be the same.

**Comment by jkotalik** [comment]
Re-adding a blocked label to this. I believe there is a fundamental issue with nethost.dll where it can't accept a path to dotnet. This will break anyone who doesn't specify DOTNET_ROOT, DOTNET_HOME, or dotnet on the path and pass in an absolute path to dotnet.

**Comment by jkotalik** [comment]
This will be mostly unblocked by https://github.com/dotnet/core-setup/pull/7020 once we update dependencies.

We will need to keep our logic to find where dotnet is if it is only available via the path. We can still feed the dotnet location to nethost.dll to find hostfxr, but nethost will not path probe.

Here is some pseudo code for how we would find dotnet now:
- Call nethost passing in either the path to the application dll and/or the path to dotnet.exe.
- If dotnet.exe isn't found, try to find dotnet on PATH.
- If we find dotnet.exe, call nethost.dll again with the newly found path to dotnet.

**Comment by dougbu** [comment]
@jkotalik will core-setup packages with version 3.0.0-preview8-27904-06 get you what you need? That was the latest when I manually updated the dependencies a couple of days ago (see #11939).

**Comment by jkotalik** [comment]
Closing for now, will reopen one day....

**Pull request closed without merging**